src/NUnitFramework/framework/Internal/Builders/NUnitTestFixtureBuilder.cs

```diff
@@ -128,10 +128,26 @@ public TestSuite BuildFrom(ITypeInfo typeInfo, ITestFixtureData testFixtureData)
             var fixture = new TestFixture(typeInfo, arguments);
 
             string name = fixture.Name;
-            if (testFixtureData.TestName != null)
-                fixture.Name = testFixtureData.TestName;
-            else if (arguments != null && arguments.Length > 0)
-                fixture.Name = typeInfo.GetDisplayName(arguments);
+
+            var parms = testFixtureData as TestParameters;
+            if (parms != null)
+            {
+                parms.ApplyToTest(fixture, _nameGenerator);
+            }
+            else
+            {
+                if (testFixtureData.TestName != null)
+                    fixture.Name = testFixtureData.TestName;
+                else if (arguments != null && arguments.Length > 0)
+                    fixture.Name = typeInfo.GetDisplayName(arguments);
+
+                if (fixture.RunState != RunState.NotRunnable)
+                    fixture.RunState = testFixtureData.RunState;
+
+                foreach (string key in testFixtureData.Properties.Keys)
+                    foreach (object val in testFixtureData.Properties[key])
+                        fixture.Properties.Add(key, val);
+            }
```

Comment:
Is there a reason setting properties/runstate was moved up to where testnames are set? Digging in to the code, I realise this is also done in `parms.ApplyToTest` - but it now looks these steps aren't done if a test has parameters? I'd rather this logic was moved out of test name setting - and duplicated for test cases if need be. (Or better yet - helper class! 😄)

Reply:
Yes. Going along with https://github.com/nunit/nunit/pull/2538#discussion_r156253485, the idea is to make the parameters responsible for looking at their private data and uniformly applying themselves whether it's a test case or a test fixture.

When a fixture doesn't have a TestParameters instance, it has to fall back to doing it itself. This is ugly. The fix I'd expect is for TestParameters to always be created for it, just like the test case builder does.

`TestParameters` *is* supposed to be the helper class. 😃 If I made sure there is always a TestParameters instance, would that address your concerns? Or is there another pattern that might be better for us than tell-don't-ask?

Reply:
I'm not sure TestParameters would be the right name for such a class, but that might be a nicer idea in general! I think such a refactor is outside of the scope of this PR now however, I didn't realises some of these points you're mentioning while reviewing. 😄